encoding/gzip.go

```diff
@@ -50,17 +51,28 @@ func (w *gzipWriter) Write(b []byte) (int, error) {
 
 	if w.WriteCloser == nil {
 		if w.canGzip() {
+			w.Header().Del("Content-Length")
 			w.Header().Set("Content-Encoding", "gzip")
 			w.Header().Add("Vary", "Accept-Encoding")
 			w.WriteCloser = gzip.NewWriter(w.ResponseWriter)
 		} else {
 			w.WriteCloser = nopCloser{w.ResponseWriter}
 		}
-	}
 
+		if w.code != 0 {
+			w.ResponseWriter.WriteHeader(w.code)
+		}
+	}
 	return w.WriteCloser.Write(b)
 }
 
+func (w *gzipWriter) WriteHeader(code int) {
+	w.Lock()
+	defer w.Unlock()
+
+	w.code = code
+}
```

Comment:
This `gzipWriter.WriteHeader` should delegate to the embedded `http.ResponseWriter`.  But if you delegate, I can see that this will prevent the headers from being mutated to support gzip from the first write.

Instead of communicating the order of `Write` and `WriteHeader` through `w.code`, factor out the setup of the headers and embedded WriteCloser in both methods, making sure to always delegate to the embedded interface.